File: generator-common/src/main/java/io/github/potjerodekool/openapi/common/generate/OpenApiWalker.java
package io.github.potjerodekool.openapi.common.generate;

import io.swagger.models.HttpMethod;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.Operation;
import io.swagger.v3.oas.models.PathItem;
import io.swagger.v3.oas.models.media.Content;
import io.swagger.v3.oas.models.media.Schema;

public final class OpenApiWalker {

    private OpenApiWalker() {
    }

    public static void walk(final OpenAPI api,
                            final OpenApiWalkerListener listener) {
        api.getPaths()
                .forEach((path, pathItem) -> visitPath(api, path, pathItem, listener));
    }

    private static void visitPath(final OpenAPI openAPI,
                                  final String path,
                                  final PathItem pathItem,
                                  final OpenApiWalkerListener listener) {
        visitOperation(openAPI, path, HttpMethod.POST, pathItem.getPost(), listener);
        visitOperation(openAPI, path, HttpMethod.GET, pathItem.getGet(), listener);
        visitOperation(openAPI, path, HttpMethod.PUT, pathItem.getPut(), listener);
        visitOperation(openAPI, path, HttpMethod.DELETE, pathItem.getDelete(), listener);
        visitOperation(openAPI, path, HttpMethod.PATCH, pathItem.getPatch(), listener);
        visitOperation(openAPI, path, HttpMethod.OPTIONS, pathItem.getOptions(), listener);
        visitOperation(openAPI, path, HttpMethod.HEAD, pathItem.getHead(), listener);
    }

    private static void visitOperation(final OpenAPI openAPI,
                                       final String path,
                                       final HttpMethod httpMethod,
                                       final Operation operation,
                                       final OpenApiWalkerListener listener) {
        if (operation != null) {
            listener.visitOperation(openAPI, httpMethod, path, operation);
            final var requestBody = operation.getRequestBody();

            if (requestBody != null) {
                final var content = requestBody.getContent();
                visitContent(openAPI, httpMethod, path, operation, content, listener);
            }

            final var responses = operation.getResponses();

            if (responses != null) {
                responses.forEach((responseCode, response) ->
                        visitContent(openAPI, httpMethod, path, operation, response.getContent(), listener));
            }
        }
    }

    private static void visitContent(final OpenAPI openAPI,
                                     final HttpMethod httpMethod,
                                     final String path,
                                     final Operation operation,
                                     final Content content,
                                     final OpenApiWalkerListener listener) {
        if (content != null) {
            final var mediaType = content.get(ContentTypes.JSON);

            if (mediaType != null) {
                final var schema = mediaType.getSchema();

                if (schema != null) {
                    visitSchema(openAPI, httpMethod, path, operation, schema, listener);
                }
            }
        }
    }

    private static void visitSchema(final OpenAPI openAPI,
                                    final HttpMethod httpMethod,
                                    final String path,
                                    final Operation operation,
                                    final Schema<?> schema,
                                    final OpenApiWalkerListener listener) {
        listener.visitSchema(openAPI, httpMethod, path, operation, schema, null);
    }

}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/generate/OpenApiWalker.kt
package io.github.potjerodekool.openapi.common.generate

import io.github.potjerodekool.openapi.common.ApiConfiguration
import io.github.potjerodekool.openapi.common.util.OpenApiUtils.getMediaType
import io.swagger.models.HttpMethod
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.Operation
import io.swagger.v3.oas.models.PathItem
import io.swagger.v3.oas.models.media.Content
import io.swagger.v3.oas.models.media.Schema
import io.swagger.v3.oas.models.responses.ApiResponse

class OpenApiWalker(
    private val api: OpenAPI,
    private val apiConfiguration: ApiConfiguration
) {
    fun walk(listener: OpenApiWalkerListener) {
        api.paths
            .forEach { path: String, pathItem: PathItem ->
                visitPath(
                    api, path, pathItem, listener
                )
            }
    }

    private fun visitPath(
        openAPI: OpenAPI,
        path: String,
        pathItem: PathItem,
        listener: OpenApiWalkerListener
    ) {
        visitOperation(openAPI, path, HttpMethod.POST, pathItem.post, listener)
        visitOperation(openAPI, path, HttpMethod.GET, pathItem.get, listener)
        visitOperation(openAPI, path, HttpMethod.PUT, pathItem.put, listener)
        visitOperation(openAPI, path, HttpMethod.DELETE, pathItem.delete, listener)
        visitOperation(openAPI, path, HttpMethod.PATCH, pathItem.patch, listener)
        visitOperation(openAPI, path, HttpMethod.OPTIONS, pathItem.options, listener)
        visitOperation(openAPI, path, HttpMethod.HEAD, pathItem.head, listener)
    }

    private fun visitOperation(
        openAPI: OpenAPI,
        path: String,
        httpMethod: HttpMethod,
        operation: Operation?,
        listener: OpenApiWalkerListener
    ) {
        if (operation != null) {
            listener.visitOperation(openAPI, httpMethod, path, operation, apiConfiguration)
            val requestBody = operation.requestBody

            if (requestBody != null) {
                val content = requestBody.content
                visitContent(openAPI, httpMethod, path, operation, content, listener)
            }

            val responses = operation.responses

            responses?.forEach { responseCode: String?, response: ApiResponse ->
                visitContent(
                    openAPI,
                    httpMethod,
                    path,
                    operation,
                    response.content,
                    listener
                )
            }
        }
    }

    private fun visitContent(
        openAPI: OpenAPI,
        httpMethod: HttpMethod,
        path: String,
        operation: Operation,
        content: Content?,
        listener: OpenApiWalkerListener
    ) {
        if (content != null) {
            listener.visitContent(
                openAPI,
                httpMethod,
                path,
                operation,
                content,
                apiConfiguration
            )

            val mt = getMediaType(content)

            if (mt != null) {
                val mediaType = mt.value
                val contentType = mt.key
                val schema = mediaType.schema

                if (schema != null) {
                    visitSchema(openAPI, httpMethod, path, operation, schema, contentType, listener)
                }
            }
        }
    }

    private fun visitSchema(
        openAPI: OpenAPI,
        httpMethod: HttpMethod,
        path: String,
        operation: Operation,
        schema: Schema<*>,
        contentType: ContentType,
        listener: OpenApiWalkerListener
    ) {
        listener.visitSchema(openAPI, httpMethod, path, operation, schema, contentType, null, apiConfiguration)
    }
}
